### PR TITLE
sketchybar: switch to apple-sdk_15

### PR DESCRIPTION
**Pull request by khaneliman**
## Things done
Bump SDK to 15, upstream uses available checks to determine functionality based on available SDK. 
ie: https://github.com/FelixKratz/SketchyBar/blob/ed1684d348fc4d4319c7f9835141524c524bacb3/src/window.c#L240-L250
>if (__builtin_available(macOS 14.0, *)) {

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by khaneliman** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 352620`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>sketchybar</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>sketchybar</li>
  </ul>
</details>

**Comment by khaneliman** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 352620`

---
### `x86_64-linux`

---
### `aarch64-linux`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>sketchybar</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>sketchybar</li>
  </ul>
</details>

**Comment by khaneliman** [comment]
Updating per new understanding of the codebase and the darwinMinVersionHook. The highest minimum available feature check is SDK 14. Setting that as the minSDK version.

**Comment by emilazy** [comment]
If upstream use availability checks why do we need to set a minimum version?

**Comment by khaneliman** [comment]
> If upstream use availability checks why do we need to set a minimum version?

Based on our other conversation, I thought the logic was basically "this should be built with the latest SDK available to get all features, but because they have code that relies on SDK 14 it can't be built by anything lower so that's the minimum we set" ?

**Comment by emilazy** [comment]
Ah, I see the confusion. Okay, let’s have a crack at explaining this again – and I do appreciate getting the opportunity to do so, because we need to figure out how to explain this to packagers!:

* The *SDK version* determines what APIs are available. `apple-sdk_n` requests the SDK for macOS *n* at minimum, which contains definitions for APIs supported by versions up to and including macOS *n*. Thus, with `apple-sdk_n` you cannot target APIs introduced in macOS *n* + 1, even based on runtime checks.

* The *deployment target* or *minimum version* determines what macOS version is *assumed*. It gets noted in executable files, and – critically – *disables* availability checks. So, for instance, if you set the minimum version to 14.0, then you are telling the compiler that you have no interest in running the executable on macOS 13; `if (@available(macOS 14, *))` will be a no‐op, and rather than the APIs you use inside such a block being weakly linked and conditioned at runtime based on the running OS version, the executable will simply fail to load on macOS 13 due to missing symbols. If you unconditionally use an API that requires a macOS version newer than your deployment target, you get a compiler warning, but it goes through (and the executable will fail to load on those old versions, unless you do your own weak linking tricks and runtime checks – MoltenVK does this).

* There is also some kind of notion of a *maximum version*, but I don’t really know why. It means that using APIs newer than that one will error out, I believe. In practice this is always set to the SDK version for us.

If you want to support building with the macOS 11 SDK, deploying down to 10.15, but optionally using APIs up to macOS 13, then you need to guard your `if (@available(macOS 13, *))` blocks with something like `#ifdef MAC_OS_VERSION_13_0`, because otherwise building with the macOS 12 SDK will fail due to the lack of definitions for those macOS 13 APIs. Sometimes this will look like `#if defined(MAC_OS_VERSION_13_0) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_VERSION_13_0` when people want to respect the maximum version thing.

tl;dr: Set the SDK so that all used APIs are available, even if they’re behind `#ifdef`/`@available`/`__builtin_available` checks; set the minimum version as low as it can go (whatever upstream says, or failing that the default if it will work, but if there are warnings about using newer APIs without an availability check then you should bump it unless you know they do their own separate runtime checks). Generally you will need to specify the SDK version more often than you need to specify the minimum runtime version.

Please do let me know if anything is still confusing, because honestly it’s fairly confusing and we need to figure out how to explain it to people.

**Comment by khaneliman** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 352620`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>sketchybar</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>sketchybar</li>
  </ul>
</details>

**Comment by khaneliman** [comment]
@emilazy this one should be good now.

**Review by emilazy — APPROVED** [review]
